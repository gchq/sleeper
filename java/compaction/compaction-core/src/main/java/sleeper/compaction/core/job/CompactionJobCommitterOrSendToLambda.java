/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.compaction.core.job.commit.CompactionCommitMessage;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.properties.table.TablePropertiesProvider;
import sleeper.core.statestore.ReplaceFileReferencesRequest;
import sleeper.core.statestore.StateStoreProvider;
import sleeper.core.statestore.commit.StateStoreCommitRequest;
import sleeper.core.statestore.transactionlog.transaction.impl.ReplaceFileReferencesTransaction;
import sleeper.core.table.TableStatus;
import sleeper.core.tracker.compaction.job.CompactionJobTracker;
import sleeper.core.tracker.compaction.job.update.CompactionJobFinishedEvent;

import java.time.Instant;
import java.util.List;
import java.util.function.Supplier;

import static sleeper.core.properties.table.TableProperty.COMPACTION_JOB_ASYNC_BATCHING;
import static sleeper.core.properties.table.TableProperty.COMPACTION_JOB_COMMIT_ASYNC;

public class CompactionJobCommitterOrSendToLambda {
    public static final Logger LOGGER = LoggerFactory.getLogger(CompactionJobCommitterOrSendToLambda.class);

    private final TablePropertiesProvider tablePropertiesProvider;
    private final StateStoreProvider stateStoreProvider;
    private final CompactionJobTracker tracker;
    private final CommitQueueSender jobCommitQueueSender;
    private final BatchedCommitQueueSender batchedCommitQueueSender;
    private final Supplier<Instant> timeSupplier;

    public CompactionJobCommitterOrSendToLambda(
            TablePropertiesProvider tablePropertiesProvider, StateStoreProvider stateStoreProvider,
            CompactionJobTracker tracker, CommitQueueSender jobCommitQueueSender, BatchedCommitQueueSender batchedCommitQueueSender) {
        this(tablePropertiesProvider, stateStoreProvider, tracker, jobCommitQueueSender, batchedCommitQueueSender, Instant::now);
    }

    public CompactionJobCommitterOrSendToLambda(
            TablePropertiesProvider tablePropertiesProvider, StateStoreProvider stateStoreProvider, CompactionJobTracker tracker,
            CommitQueueSender jobCommitQueueSender, BatchedCommitQueueSender batchedCommitQueueSender, Supplier<Instant> timeSupplier) {
        this.tablePropertiesProvider = tablePropertiesProvider;
        this.stateStoreProvider = stateStoreProvider;
        this.tracker = tracker;
        this.jobCommitQueueSender = jobCommitQueueSender;
        this.batchedCommitQueueSender = batchedCommitQueueSender;
        this.timeSupplier = timeSupplier;
    }

    public void commit(CompactionJob job, CompactionJobFinishedEvent finishedEvent) {
        TableProperties tableProperties = tablePropertiesProvider.getById(job.getTableId());
        TableStatus table = tableProperties.getStatus();
        tracker.jobFinished(finishedEvent);
        ReplaceFileReferencesRequest.Builder requestBuilder = job.replaceFileReferencesRequestBuilder(finishedEvent.getRecordsProcessed().getRecordsWritten());
        if (tableProperties.getBoolean(COMPACTION_JOB_COMMIT_ASYNC)) {
            ReplaceFileReferencesRequest request = requestBuilder.taskId(finishedEvent.getTaskId()).jobRunId(finishedEvent.getJobRunId()).build();
            if (tableProperties.getBoolean(COMPACTION_JOB_ASYNC_BATCHING)) {
                CompactionCommitMessage message = new CompactionCommitMessage(table.getTableUniqueId(), request);
                LOGGER.debug("Sending asynchronous request to commit batcher: {}", message);
                batchedCommitQueueSender.send(message);
                LOGGER.info("Sent compaction job {} to batcher queue to be committed asynchronously to table {}", job.getId(), table);
            } else {
                StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create(table.getTableUniqueId(),
                        new ReplaceFileReferencesTransaction(List.of(request)));
                LOGGER.debug("Sending asynchronous request to state store committer: {}", commitRequest);
                jobCommitQueueSender.send(commitRequest);
                LOGGER.info("Sent compaction job {} to state store queue to be committed asynchronously to table {}", job.getId(), table);
            }
        } else {
            LOGGER.debug("Committing compaction job {} inside compaction task", job.getId());
            new ReplaceFileReferencesTransaction(List.of(requestBuilder.build()))
                    .synchronousCommit(stateStoreProvider.getStateStore(tableProperties));
            tracker.jobCommitted(job.committedEventBuilder(timeSupplier.get())
                    .jobRunId(finishedEvent.getJobRunId())
                    .taskId(finishedEvent.getTaskId())
                    .build());
            LOGGER.info("Successfully committed compaction job {} to table {}", job.getId(), table);
        }
    }

    public interface CommitQueueSender {
        void send(StateStoreCommitRequest commitRequest);
    }

    public interface BatchedCommitQueueSender {
        void send(CompactionCommitMessage commitRequest);
    }
}
