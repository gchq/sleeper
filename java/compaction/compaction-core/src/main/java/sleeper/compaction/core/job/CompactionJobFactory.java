/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.FileReference;
import sleeper.core.table.TableFilePaths;

import java.util.List;
import java.util.UUID;
import java.util.function.Supplier;

import static java.util.stream.Collectors.toList;
import static sleeper.core.properties.table.TableProperty.ITERATOR_CLASS_NAME;
import static sleeper.core.properties.table.TableProperty.ITERATOR_CONFIG;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;

public class CompactionJobFactory {

    private final String tableId;
    private final TableFilePaths outputFilePaths;
    private final String iteratorClassName;
    private final String iteratorConfig;
    private final Supplier<String> jobIdSupplier;

    public CompactionJobFactory(InstanceProperties instanceProperties, TableProperties tableProperties) {
        this(instanceProperties, tableProperties, () -> UUID.randomUUID().toString());
    }

    public CompactionJobFactory(InstanceProperties instanceProperties, TableProperties tableProperties, Supplier<String> jobIdSupplier) {
        tableId = tableProperties.get(TABLE_ID);
        outputFilePaths = TableFilePaths.buildDataFilePathPrefix(instanceProperties, tableProperties);
        iteratorClassName = tableProperties.get(ITERATOR_CLASS_NAME);
        iteratorConfig = tableProperties.get(ITERATOR_CONFIG);
        this.jobIdSupplier = jobIdSupplier;
    }

    public String getOutputFilePrefix() {
        return outputFilePaths.getFilePathPrefix();
    }

    public CompactionJob createCompactionJob(
            List<FileReference> files, String partition) {
        return createCompactionJob(jobIdSupplier.get(), files, partition);
    }

    public CompactionJob createCompactionJob(
            String jobId, List<FileReference> files, String partition) {
        for (FileReference fileReference : files) {
            if (!partition.equals(fileReference.getPartitionId())) {
                throw new IllegalArgumentException("Found file with partition which is different to the provided partition (partition = "
                        + partition + ", FileReference = " + fileReference);
            }
        }
        return createCompactionJobWithFilenames(jobId,
                files.stream().map(FileReference::getFilename).collect(toList()),
                partition);
    }

    public CompactionJob createCompactionJobWithFilenames(
            String jobId, List<String> filenames, String partitionId) {
        String outputFile = outputFilePaths.constructPartitionParquetFilePath(partitionId, jobId);
        return CompactionJob.builder()
                .tableId(tableId)
                .jobId(jobId)
                .inputFiles(filenames)
                .outputFile(outputFile)
                .partitionId(partitionId)
                .iteratorClassName(iteratorClassName)
                .iteratorConfig(iteratorConfig)
                .build();
    }
}
