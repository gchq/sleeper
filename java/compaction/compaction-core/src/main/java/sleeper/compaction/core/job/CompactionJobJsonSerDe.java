/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.lang.reflect.Type;

/**
 * A GSON plugin to serialise/deserialise a compaction job.
 */
public class CompactionJobJsonSerDe implements JsonSerializer<CompactionJob>, JsonDeserializer<CompactionJob> {
    @Override
    public CompactionJob deserialize(JsonElement element, Type type, JsonDeserializationContext context) throws JsonParseException {
        try {
            return CompactionJobSerDe.deserialiseFromString(element.getAsString());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public JsonElement serialize(CompactionJob job, Type type, JsonSerializationContext context) {
        try {
            return new JsonPrimitive(CompactionJobSerDe.serialiseToString(job));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
