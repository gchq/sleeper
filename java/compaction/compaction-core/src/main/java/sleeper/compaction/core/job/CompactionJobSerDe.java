/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import org.apache.commons.codec.binary.Base64;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Serialises and deserialises a compaction job to and from a JSON string.
 */
public class CompactionJobSerDe {

    private CompactionJobSerDe() {
    }

    public static String serialiseToString(CompactionJob compactionJob) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        dos.writeUTF(compactionJob.getTableId());
        dos.writeUTF(compactionJob.getId());
        dos.writeUTF(compactionJob.getPartitionId());
        dos.writeInt(compactionJob.getInputFiles().size());
        for (String inputFile : compactionJob.getInputFiles()) {
            dos.writeUTF(inputFile);
        }
        if (null == compactionJob.getIteratorClassName()) {
            dos.writeBoolean(true);
        } else {
            dos.writeBoolean(false);
            dos.writeUTF(compactionJob.getIteratorClassName());
        }
        if (null == compactionJob.getIteratorConfig()) {
            dos.writeBoolean(true);
        } else {
            dos.writeBoolean(false);
            dos.writeUTF(compactionJob.getIteratorConfig());
        }
        dos.writeUTF(compactionJob.getOutputFile());
        dos.close();

        return Base64.encodeBase64String(baos.toByteArray());
    }

    public static CompactionJob deserialiseFromString(String serialisedJob) throws IOException {
        byte[] bytes = Base64.decodeBase64(serialisedJob);
        ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
        DataInputStream dis = new DataInputStream(bais);
        String tableId = dis.readUTF();
        CompactionJob.Builder compactionJobBuilder = CompactionJob.builder()
                .tableId(tableId)
                .jobId(dis.readUTF())
                .partitionId(dis.readUTF());
        int numInputFiles = dis.readInt();
        List<String> inputFiles = new ArrayList<>(numInputFiles);
        for (int i = 0; i < numInputFiles; i++) {
            inputFiles.add(dis.readUTF());
        }
        compactionJobBuilder.inputFiles(inputFiles)
                .iteratorClassName(!dis.readBoolean() ? dis.readUTF() : null)
                .iteratorConfig(!dis.readBoolean() ? dis.readUTF() : null);
        compactionJobBuilder.outputFile(dis.readUTF());
        dis.close();
        return compactionJobBuilder.build();
    }
}
