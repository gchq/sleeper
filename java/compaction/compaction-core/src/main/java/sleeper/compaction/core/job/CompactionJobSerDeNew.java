/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import sleeper.core.util.GsonConfig;

public class CompactionJobSerDeNew {

    private final Gson gson;
    private final Gson gsonPrettyPrint;

    public CompactionJobSerDeNew() {
        GsonBuilder builder = GsonConfig.standardBuilder();
        gson = builder.create();
        gsonPrettyPrint = builder.setPrettyPrinting().create();
    }

    public String toJsonPrettyPrint(CompactionJob job) {
        return gsonPrettyPrint.toJson(job);
    }

    public String toJson(CompactionJob job) {
        return gson.toJson(job);
    }

    public CompactionJob fromJson(String json) {
        return gson.fromJson(json, CompactionJob.class);
    }

}
