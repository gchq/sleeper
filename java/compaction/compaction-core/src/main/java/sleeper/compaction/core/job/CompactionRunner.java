/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

import sleeper.core.iterator.IteratorCreationException;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.range.Region;
import sleeper.core.tracker.job.run.RowsProcessed;

import java.io.IOException;

/**
 * An implementation of compaction, to take a number of sorted input files and merge them into one fully sorted file.
 */
@FunctionalInterface
public interface CompactionRunner {

    /**
     * Compacts the input files of a compaction job into one output file.
     *
     * @param  job                       the job
     * @param  tableProperties           the configuration of the Sleeper table this takes place in
     * @param  region                    the region to be read for the compaction (usually the region of the partition
     *                                   it takes place in)
     * @return                           a report of the number of rows processed
     * @throws IOException               a failure reading or writing files
     * @throws IteratorCreationException a problem creating any configured iterators
     */
    RowsProcessed compact(CompactionJob job, TableProperties tableProperties, Region region) throws IOException, IteratorCreationException;
}
