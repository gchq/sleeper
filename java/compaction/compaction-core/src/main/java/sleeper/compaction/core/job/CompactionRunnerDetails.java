/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job;

public interface CompactionRunnerDetails {
    /**
     * Some compaction implementations may use hardware acceleration such as GPUs.
     *
     * @return true iff this compaction implementation uses any sort of hardware acceleration
     */
    default boolean isHardwareAccelerated() {
        return false;
    }

    /**
     * What language is this implemented in? If multiple languages are used, the primary
     * one used for performing the compaction computation should be returned.
     *
     * @return the principal implementation language for this compactor
     */
    default String implementationLanguage() {
        return "Java";
    }
}
