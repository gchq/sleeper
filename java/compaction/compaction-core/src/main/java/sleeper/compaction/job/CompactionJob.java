/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.job;

import java.util.List;
import java.util.Objects;

/**
 * Contains the definition of a compaction job. This includes the ID of the job,
 * a list of the input files, the output file, and the ID of the partition.
 * <p>
 * This should fully define the job to be performed, so that no further queries of
 * the state store should be required in order to start it.
 */
public class CompactionJob {
    private final String tableId;
    private final String jobId;
    private final List<String> inputFiles;
    private final String outputFile;
    private final String partitionId;
    private final String iteratorClassName;
    private final String iteratorConfig;

    private CompactionJob(Builder builder) {
        tableId = Objects.requireNonNull(builder.tableId, "tableId must not be null");
        jobId = Objects.requireNonNull(builder.jobId, "jobId must not be null");
        inputFiles = Objects.requireNonNull(builder.inputFiles, "inputFiles must not be null");
        outputFile = Objects.requireNonNull(builder.outputFile, "outputFile must not be null");
        partitionId = Objects.requireNonNull(builder.partitionId, "partitionId must not be null");
        iteratorClassName = builder.iteratorClassName;
        iteratorConfig = builder.iteratorConfig;
        checkDuplicates(inputFiles);
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getTableId() {
        return tableId;
    }

    public String getId() {
        return jobId;
    }

    public String getPartitionId() {
        return partitionId;
    }

    public List<String> getInputFiles() {
        return inputFiles;
    }

    /**
     * Checks that there are no duplicate entries present in the list of files.
     *
     * @param  <T>                      generic type of list
     * @param  files                    list of entries to check
     * @throws IllegalArgumentException if there are any duplicate entries
     */
    private static <T> void checkDuplicates(List<T> files) {
        if (files.stream()
                .distinct()
                .count() != files.size()) {
            throw new IllegalArgumentException("Duplicate entry present in file list");
        }
    }

    public String getOutputFile() {
        return outputFile;
    }

    public String getIteratorClassName() {
        return iteratorClassName;
    }

    public String getIteratorConfig() {
        return iteratorConfig;
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }
        if (object == null || getClass() != object.getClass()) {
            return false;
        }
        CompactionJob that = (CompactionJob) object;
        return Objects.equals(tableId, that.tableId) && Objects.equals(jobId, that.jobId)
                && Objects.equals(inputFiles, that.inputFiles) && Objects.equals(outputFile, that.outputFile)
                && Objects.equals(partitionId, that.partitionId) && Objects.equals(iteratorClassName, that.iteratorClassName)
                && Objects.equals(iteratorConfig, that.iteratorConfig);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tableId, jobId, inputFiles, outputFile, partitionId,
                iteratorClassName, iteratorConfig);
    }

    @Override
    public String toString() {
        return "CompactionJob{" +
                "tableId='" + tableId + '\'' +
                ", jobId='" + jobId + '\'' +
                ", inputFiles=" + inputFiles +
                ", outputFile='" + outputFile + '\'' +
                ", partitionId='" + partitionId + '\'' +
                ", iteratorClassName='" + iteratorClassName + '\'' +
                ", iteratorConfig='" + iteratorConfig + '\'' +
                '}';
    }

    public static final class Builder {
        private String tableId;
        private String jobId;
        private List<String> inputFiles;
        private String outputFile;
        private String partitionId;
        private String iteratorClassName;
        private String iteratorConfig;

        private Builder() {
        }

        public Builder tableId(String tableId) {
            this.tableId = tableId;
            return this;
        }

        public Builder jobId(String jobId) {
            this.jobId = jobId;
            return this;
        }

        public Builder inputFiles(List<String> inputFiles) {
            this.inputFiles = inputFiles;
            return this;
        }

        public Builder outputFile(String outputFile) {
            this.outputFile = outputFile;
            return this;
        }

        public Builder partitionId(String partitionId) {
            this.partitionId = partitionId;
            return this;
        }

        public Builder iteratorClassName(String iteratorClassName) {
            this.iteratorClassName = iteratorClassName;
            return this;
        }

        public Builder iteratorConfig(String iteratorConfig) {
            this.iteratorConfig = iteratorConfig;
            return this;
        }

        public CompactionJob build() {
            return new CompactionJob(this);
        }
    }
}
