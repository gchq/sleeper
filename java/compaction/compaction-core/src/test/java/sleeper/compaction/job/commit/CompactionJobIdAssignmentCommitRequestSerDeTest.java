/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.job.commit;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.compaction.job.CompactionJob;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static sleeper.compaction.job.commit.CompactionJobIdAssignmentCommitRequestTestHelper.requestToAssignFilesToJobs;

public class CompactionJobIdAssignmentCommitRequestSerDeTest {

    private final CompactionJobIdAssignmentCommitRequestSerDe serDe = new CompactionJobIdAssignmentCommitRequestSerDe();

    @Test
    void shouldSerialiseCompactionJobIdAssignmentCommitRequest() throws Exception {
        // Given
        CompactionJob job1 = CompactionJob.builder()
                .tableId("test-table")
                .jobId("test-job-1")
                .inputFiles(List.of("file1.parquet", "file2.parquet"))
                .outputFile("test-output-1.parquet")
                .partitionId("test-partition-id")
                .build();
        CompactionJob job2 = CompactionJob.builder()
                .tableId("test-table")
                .jobId("test-job-2")
                .inputFiles(List.of("file3.parquet", "file4.parquet"))
                .outputFile("test-output-2.parquet")
                .partitionId("test-partition-id")
                .build();
        CompactionJobIdAssignmentCommitRequest commit = requestToAssignFilesToJobs(List.of(job1, job2), "test-table");

        // When
        String json = serDe.toJsonPrettyPrint(commit);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commit);
        Approvals.verify(json);
    }

    @Test
    void shouldFailToDeserialiseNonCompactionJobIdAssignmentCommitRequest() {
        assertThatThrownBy(() -> serDe.fromJson("{\"type\": \"OTHER\", \"request\":{}}"))
                .isInstanceOf(IllegalArgumentException.class);
    }
}
