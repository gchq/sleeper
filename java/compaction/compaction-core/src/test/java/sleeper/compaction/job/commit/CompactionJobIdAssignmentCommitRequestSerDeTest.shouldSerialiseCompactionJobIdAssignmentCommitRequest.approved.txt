{
  "type": "COMPACTION_JOB_ID_ASSIGNMENT",
  "request": {
    "tableId": "test-table",
    "assignJobIdRequests": [
      {
        "jobId": "test-job-1",
        "partitionId": "test-partition-id",
        "filenames": [
          "file1.parquet",
          "file2.parquet"
        ]
      },
      {
        "jobId": "test-job-2",
        "partitionId": "test-partition-id",
        "filenames": [
          "file3.parquet",
          "file4.parquet"
        ]
      }
    ]
  }
}